pkg/network/node/egressip.go

```diff
@@ -0,0 +1,300 @@
+
+	"github.com/vishvananda/netlink"
+)
+
+type nodeEgress struct {
+	nodeIP    string
+	egressIPs sets.String
+}
+
+type namespaceEgress struct {
+	vnid uint32
+
+	// claimedIP is the egress IP it wants (NetNamespace.EgressIP[0]), or "" for none
+	claimedIP string
+	// assignedIP is an egress IP actually in use on nodeIP
+	assignedIP string
+	nodeIP     string
+}
+
+type egressIPWatcher struct {
+	sync.Mutex
+
+	localIP string
+	oc      *ovsController
+
+	networkClient networkclient.Interface
+	iptables      *NodeIPTables
+
+	// from HostSubnets
+	nodesByNodeIP   map[string]*nodeEgress
+	nodesByEgressIP map[string]*nodeEgress
+
+	// From NetNamespaces
+	namespacesByVNID     map[uint32]*namespaceEgress
+	namespacesByEgressIP map[string]*namespaceEgress
+
+	localEgressLink      netlink.Link
+	localEgressIPMaskLen int
+
+	testModeChan chan string
+}
+
+func newEgressIPWatcher(localIP string, oc *ovsController) *egressIPWatcher {
+	return &egressIPWatcher{
+		localIP: localIP,
+		oc:      oc,
+
+		nodesByNodeIP:   make(map[string]*nodeEgress),
+		nodesByEgressIP: make(map[string]*nodeEgress),
+
+		namespacesByVNID:     make(map[uint32]*namespaceEgress),
+		namespacesByEgressIP: make(map[string]*namespaceEgress),
+	}
+}
+
+func (eip *egressIPWatcher) Start(networkClient networkclient.Interface, iptables *NodeIPTables) error {
+	eip.iptables = iptables
+	eip.networkClient = networkClient
+
+	go utilwait.Forever(eip.watchHostSubnets, 0)
+	go utilwait.Forever(eip.watchNetNamespaces, 0)
+	return nil
+}
+
+func ipToHex(ip string) string {
+	bytes := net.ParseIP(ip)
+	if bytes == nil {
+		return "invalid IP: shouldn't happen"
+	}
+	bytes = bytes.To4()
+	return fmt.Sprintf("0x%02x%02x%02x%02x", bytes[0], bytes[1], bytes[2], bytes[3])
+}
+
+func (eip *egressIPWatcher) watchHostSubnets() {
+	common.RunEventQueue(eip.networkClient.Network().RESTClient(), common.HostSubnets, func(delta cache.Delta) error {
+		hs := delta.Object.(*networkapi.HostSubnet)
+
+		var egressIPs []string
+		if delta.Type != cache.Deleted {
+			egressIPs = hs.EgressIPs
+		}
+
+		eip.updateNode(hs.HostIP, egressIPs)
+		return nil
+	})
+}
+
+func (eip *egressIPWatcher) updateNode(nodeIP string, nodeEgressIPs []string) {
+	eip.Lock()
+	defer eip.Unlock()
+
+	node := eip.nodesByNodeIP[nodeIP]
+	if node == nil {
+		if len(nodeEgressIPs) == 0 {
+			return
+		}
+		node = &nodeEgress{nodeIP: nodeIP, egressIPs: sets.NewString()}
+		eip.nodesByNodeIP[nodeIP] = node
+	} else if len(nodeEgressIPs) == 0 {
+		delete(eip.nodesByNodeIP, nodeIP)
+	}
+	oldEgressIPs := node.egressIPs
+	node.egressIPs = sets.NewString(nodeEgressIPs...)
+
+	// Process new EgressIPs
+	for _, ip := range node.egressIPs.Difference(oldEgressIPs).UnsortedList() {
+		eip.nodesByEgressIP[ip] = node
+		hex := ipToHex(ip)
+
+		if nodeIP == eip.localIP {
+			if err := eip.claimEgressIP(ip, hex); err != nil {
+				glog.Errorf("Error claiming Egress IP %q: %v", ip, err)
```

Comment:
Should we continue at this point? Or update the rules anyway?

Reply:
We should update the rules, but we should update them to drop the packets (so the code is currently wrong)

Reply:
(fixed)